### ISSUE TITLE
Modelica.Mechanics.MultiBody.Joints.GearConstraint

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#951]
**Reported by mtiller on 3 Jan 2013 16:04 UTC**
I don't think this model is correct.  About a year ago, Martin Malmheden and I created a test case for this but I no longer have access to it.

The issue, as I recall, was due to the fact that the gear element was implicitly grounded.  I don't know what the fix for this is, but the implicit connection to rotational ground doesn't look right here to me.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/951

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 30 Mar 2013 18:04 UTC**
Added the test model above as attachment to this ticket (was sent by Martin Malmheden to me). I just checked, and the behavior is correct: cyl2 is not locked and the angular velocity is the same as inertia2 (inertia2.w = cyl2.frame_a.R.w[1]).

Here is the reasoning for the GearConstraint model: Assume that the IdealGear is not present. Then the model describes the essential part of the gear: Two revolute joints rotate along the same axis, so frame_a and frame_b are rotating along the same axis, but with different speeds. Frame "bearing" is the housing. Until now everything is correct. In order to construct now a "real" gear, the following constraint equations must be added that describe the kinematic behavior of the gear:

  frame_a.w\*e_a   = ratio\*frame_b.w\*e_b
  frame_b.tau\*e_b = ratio\*frame_a.tau\*e_a

that is, the projection of the angular velocity of frame_a on the rotation axis of revolute joint a is proportional to the projection of the angular velocity of the frame_b on the rotation axis of revolute joint b and the projected torques are inverse proportional.

Now, the projection of the "movement" of frame_a on the revolute joint axis is exactly the axis angle (axis.phi). Therefore, the equations above are identical to (now on position and not on speed level):

  revolute_a.axis.phi = ratio\*revolute_b.axis.phi
  revolute_b.axis.tau = ratio\*revolute_a.axis.tau

and these equations in turn are exactly the equations of an IdealGear where the housing is fixed on ground! Therefore, either the equations above are introduced in the text layer, or (as done in MSL), an IdealGear with a fixed housing is connected between the two axes of the revolute joints.

To summarize, I cannot see an error in this element.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 30 Mar 2013 18:06 UTC**
Sorry, I made a small error in the description: Of course the two revolute joints can rotate along different axis (e.g. to model a bevel gear). All  the argument still hold.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 13 Nov 2014 09:42 UTC**
I'd like to see this case discussed again. The attachment:GearConstraint_check_bearing.mo and also the sample model Modelica.Mechanics.MultiBody.Examples.Rotational3DEffects.GearConstraint is an insufficient test as all support torques (coming from the torque source or the gear constraint) are grounded anyway. So there is nothing to prove here because nothing can't go wrong.

In attachment:T951.mo I copied both submodels from Modelica.Mechanics.MultiBody.Examples.Rotational3DEffects.GearConstraint and attached each to a freely moving MultiBody structure. Now inertia2.w and cyl2.frame_a.R.w![1] are different which illustrates the incorrect modeling because
 * The gearConstraint.idealGear of the top submodel is implicitly grounded. (This is the original issue.)
 * Both inertia1 and inertia2 of the bottom submodel are not mounted to any MultiBody frame. (I kept this by purpose when copying from the MultiBody sample Rotational3DEffects.GearConstraint as I do not understand why purely rotational inertias are used in this sample model.)

There should be fix that both submodels of attachment:T951.mo give identical results. I have strong interest to see this fixed soon.

**Comment by modelica-trac-importer** [comment]
**Comment by tobolar on 14 Nov 2014 16:16 UTC**
To couple Multibody with 1D is always a little bit tricky.

I have changed and added some elements (see attachment
[T951_fixed.mo](https://trac.modelica.org/Modelica/attachment/ticket/951/T951_fixed.mo)) so that there are comparable results from sensors absoluteSensor_1D and absoluteSensor_MBS (see angles and position vectors r).

Therefore, I would say the model works correctly.

Nevertheless, I recommend to enhance the example Modelica.Mechanics.MultiBody.Examples.Rotational3DEffects.GearConstraint so that a user can better understand how to use the GearConstraint.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 20 Nov 2014 08:39 UTC**
Thanks for the fix. Some was obvious (e.g. Rotor1D), some was not (3D Cylinders). I can see the comparable results, too.

However I still believe it is not correct. If I replace the IdealGear by a BevelGear1D results are still comparable. But in #1606 I gave reasons why I consider BevelGear1D as incorrect. Please see also #1606 and attachment:T951_fixed_T1606.mo.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 3 Jan 2015 19:36 UTC**
Clarified in 8d1e8b88d93d28bd02720e0f2092317e95170a34 (in trunk):
- The GearConstraint model is correct (and was not modified)
- A parameter "checkTotalPower" was added (default = false). Once enabled, the total energy flow is computed flowing into this component via its 3 frames (and the result should be zero).
- "Implementation notes" added in the text layer of the GearConstraint model.
- The example model in MultiBody.Examples and the two test models in ModelicaTest in which the GearConstraint model was used, have been modified, to check the total energy flow (which happens to be close to zero, as it should be).
- Two new test models have been added (ModelicaTest.MultiBody.Parts.Rotor1D.GearConstraint3/GearConstraint4), where the mounting frame is moving. The GearConstraint4 test model is the one from #951#comment:4. The total energy flow of the GearConstraint models in these tests are close to zero, as they should be.

For your convenience, find below the implementation notes from the model. After this thorough analysis and the various test models, I am completely convinced that this component is correct. Note, also the BevelGear1D model is correct, as explained in #1606 (the energy flow computation is a bit more complicated, because the absolute and not the relative angular velocities have to be taken into account). Please re-open this ticket only, if you have a counter example, where the model clearly computes the wrong result.

Implementation notes (from the text layer of GearConstraint):

The GearConstraint model consists primarily of two revolute joints that are     connected together and to a support/mounting. In this first phase the two revolute joints can be rotated independently from each other and therefore there are two degrees of freedom. If the rotational angles of these joints would be used as generalized coordinates phi_a, phi_b with associated generalized torques tau_a, tau_b torques along the axes of rotations), then the equations of motion (Kanes' equations or Lagranges' equations of the second kind) are in the rows for phi_a, phi_b:

```
.... = ... + {...., tau_a, tau_b, ....}
```
Now the kinematic constraint for the gear is added:
```
0 = phi_a - ratio*phi_b;
```
or on velocity level:
```
0 = G * {der(phi_a), der(phi_b)};   G = [1, -ratio]
```
According to Lagranges' equations of the first kind, the generalized forces must be replaced by G'\*lambda, where lambda is the new constraint force due this constraint. Therefore, the equations of motions are changed to
```
.... = .... + {...., G'*lambda, .....}
```
This is equivalent to add the equations
```
tau_a = lambda
tau_b = -ratio*lambda
```
or
```
0 = tau_b + ratio*tau_a;
```
The two equations
```
0 = phi_a - ratio*phi_b
0 = tau_b + ratio*tau_a
```
are completely identical to the equations of an ideal gear (without mounting) that connects the axis flanges of the two revolute joints. This in turn means that the two rotational flanges of the revolute joints just have to be connected by an IdealGear component (without mounting).

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 3 Jan 2015 19:38 UTC**

**Comment by modelica-trac-importer** [comment]
**Changelog modified by otter on 3 Jan 2015 19:38 UTC**
GearConstraint: Added optional computation for the total power (should be zero), added implementation notes and more test examples

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 3 Jan 2015 19:48 UTC**
